feat: Support multi cursor line moves

2mas opened the pull request:
Fixes #68694 moving multiple consequitive lines

2mas committed:
Fixes #68694 moving multiple consequitive lines

2mas committed:
Null-check + only apply setSelections if there are multiple new selections, otherwise its already selected

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=68695) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=68695) before we can accept your contribution.<br/><br/>:x: 2mas [sign now](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=68695)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vscode?pullRequest=68695) it.</sub>

2mas commented:
Needs some more work, lines get scrambled

2mas committed:
correction, reversing selections back when moving lines down

2mas commented:
Perhaps a comment about why selections needs to be reversed and executed individually should be added for clarity, tell me if I can assist more

2mas committed:
Fixes having multiple selections on same line and moving them
Correcting multiple move-lines when having made selections in reversed order

2mas commented:
Latest commit covers scenario when selection-direction goes upwards. Reversing the processing of selections should only occur when lines are moved in the same direction as the selections are made.

Also had to make the selections distinct by line. Keeping the selections can probably be solved too since their column-positions wont be changed. If you're interested in this approach in this PR I can give that a try aswell

![temp](https://user-images.githubusercontent.com/372362/53547479-72616d00-3b2f-11e9-953d-7565d9e6abf6.gif)

2mas committed:
Stashing all selections before proceeding with selections distinct by line. Applies stash after line movement to keep previous
selections.

Aborting when moving multiple lines and the line is currently at top or end of document

2mas commented:
Decided to implement the stashing of selections anyway, also added a block when top/end of doc is reached. I'm probably done by now. :)

2mas committed:
fix strict-null-check

2mas commented:
@rebornix any status on this? Just curious

rebornix commented:
@2mas let's shoot for next week (beginning of next iteration). BTW, I'd love to see some tests as well but I can do that if the time doesn't work for you.

2mas commented:
Awesome, I’ll see if I can come up with something (limited available time I’m afraid).

2mas committed:
Adding testcases

2mas commented:
Added a few testcases and resolved a conflict with master, should be good to test-drive now.

vivodi commented:
Is this still relevant?

2mas commented:
> Is this still relevant?

It seems so, the issue is still there in the editor and keeps annoying users apparently. - https://github.com/microsoft/vscode/issues/11809#issuecomment-2531210408

No idea about the state of the code in my PR today though, spent some time on it years ago to make it good to testdrive and merge but it could have fallen short in prioritization since its probably not impacting the majority of the users, or there is something else about the solution that stops it

rebornix commented:
This change requires comprehensive testing before merging. At the same time I'm not sure if we are investing more in the textmate indentation rules. I'll leave this to @aiday-mar to decide.

aiday-mar commented on src/vs/editor/contrib/linesOperations/linesOperations.ts:
```diff
@@ -112,18 +112,74 @@ abstract class AbstractMoveLinesAction extends EditorAction {
+
+			editor.pushUndoStop();
+		} else {
+			editor.pushUndoStop();
```
Could we unify this part of the code between the two cases? We could construct the commands array entirely first and then call `executeCommands` on this `commands` array. Then after this, if we have moved multiple lines, we can set the selections.

2mas replied:
This didn't work out for me. For some reason it just doesn't process all of the commands

aiday-mar commented on src/vs/editor/contrib/linesOperations/linesOperations.ts:
```diff
@@ -112,18 +112,74 @@ abstract class AbstractMoveLinesAction extends EditorAction {
+					newSelections.push(new Selection(newStartLine, startCol, newEndLine, endCol));
+				}
+
+				// Work only with one selection per line
```
Let's put this filter operation before the for loop above, so as not to loop over uneccessary selections in the for loop. 

Also instead of doing a filter operation followed by a map operation followed by an indexOf operation which is O(n^2) where `n` is the number of selections could we just loop once over the selections, keep track of the `endLineNumber` of the last processed selection and if the current selection does not match this `endLineNumber` add the current selection to a pre-created selections array. This will allow us to reduce the complexity to O(n).

aiday-mar commented on src/vs/editor/contrib/linesOperations/linesOperations.ts:
```diff
@@ -112,18 +112,74 @@ abstract class AbstractMoveLinesAction extends EditorAction {
+					return arr.map(sel => sel['endLineNumber']).indexOf(s['endLineNumber']) === idx;
+				});
+
+				selectionDirectionDown = selections[0].endLineNumber < selections[selections.length - 1].endLineNumber;
```
I am not sure I understand this piece of code. Why do we calculate the selectionDirectionDown. What does it mean?

2mas replied:
Yes, if I do this:
![image](https://github.com/user-attachments/assets/3163d7d7-0fb5-48be-b5e1-a7ad5a3e7096)

And select some lines with `Add Previous Occurrence`:
![image](https://github.com/user-attachments/assets/7d77fbc2-ff80-4b0e-930a-77a75dac3917)

And move down:
![image](https://github.com/user-attachments/assets/15d8076c-a8eb-4081-a495-dae982d59358)


vs

![image](https://github.com/user-attachments/assets/c418d4b8-82a9-4e59-ac2b-b2603982bf06)

`Add Previous Occurrence`:
![image](https://github.com/user-attachments/assets/7d77fbc2-ff80-4b0e-930a-77a75dac3917)

move down:
![image](https://github.com/user-attachments/assets/a46f698c-eb1a-46a8-9bee-3adab014d86c)

2mas replied:
It's because "line 3" gets processed first -> moved to row 6, then the next selection is processed on row 6, which is now "line 3" -> moved to row 7, next selection gets processed on row 7, which is "line 3" -> ends up on row 8. So this is why process order must be right :)

aiday-mar reviewed (changes requested):
Hi thank you for making this PR and apologies for the delay in reviewing this PR. I think this PR would be a very valuable contribution. I have looked at it, and I think the approach is overall correct but I have several questions as well as changes I think we should make before merging. It would be good to also test this PR on many different corner cases to make sure it does not introduce bugs. I will do that too. There is also a merge conflict which needs to be resolved first.

2mas commented:
> Hi thank you for making this PR and apologies for the delay in reviewing this PR. I think this PR would be a very valuable contribution. I have looked at it, and I think the approach is overall correct but I have several questions as well as changes I think we should make before merging. It would be good to also test this PR on many different corner cases to make sure it does not introduce bugs. I will do that too. There is also a merge conflict which needs to be resolved first.

Hi! Nice to see that you find this useful, let me get back with resolved conflicts and some adjustments and clarifications. 

I have some issues running this project locally though, but I managed to get the web-editor working at least, but tests are only running half-way for some reason, perhaps I can rely on CI anyway, I'll do my best effort

2mas commented:
> What issues are you seeing while running this project locally? You should be able to compile as well as run the tests and test with OSS

Firstly tried running test.sh and code.sh in WSL when cloning into devcontainer which failed, can't remember what was logged. I got closest to full experience with running directly on Windows with node v22.15.1, python 3.13.0 and VS 2022 + workload and individual components installed as per docs, I get these types of issues when running locally in Windows:

- npm install
- npm run watch

`.\scripts\code.bat` actually worked on some more attempts now, seems to be some intermittent error previously
`.\scripts\test.bat` doesn't seem to run all tests, it prints this at the end:
```
Found --crash-reporter-directory argument. Setting crashDumps directory to be 'C:\Projects\vscode\.build\crashes'
[20020:0610/140701.484:INFO:CONSOLE(2)] "The vm module of Node.js is unsupported in Electron's renderer process due to incompatibilities with the Blink rendering engine. Crashes are likely and avoiding the module is highly recommended. This module may be removed in a future release.", source: node:electron/js2c/renderer_init (2)
[20020:0610/140701.797:INFO:CONSOLE(2)] "Platform browser has already been set. Overwriting the platform with [object Object].", source: C:\Projects\vscode\node_modules\@vscode\vscode-languagedetection\dist\lib\index.js (2)
(node:30204) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `Code - OSS --trace-deprecation ...` to show where the warning was created)
(node:30204) [DEP0025] DeprecationWarning: sys is deprecated. Use util instead.
(node:30204) ExperimentalWarning: WASI is an experimental feature and might change at any time

...

322 passing (2s)
  2 pending
  2 failing

...

2) "after each" hook for "English File menu renders mnemonics":
      + expected - actual

      -false
      +true
  AssertionError [ERR_ASSERTION]: Error: Unexpected console output in test run. Please ensure no console.[log|error|info|warn] usage in tests or runtime errors.
      at Context.<anonymous> (renderer.js:291:11)

errorlevel: 1
```

2mas committed:
fix: move out distinct line filter out of loop + sort by line

2mas committed:
fix: improve line movement logic for multiple selections and add tests, also handle order of selections

2mas committed:
fix: update suite name for moving multiple lines and refactor action execution

2mas commented:
I have pushed some updates and covered some edge cases as well, for example if one puts cursors on multiple lines with no order, we will still sort them and apply the move only once per line.

There are some conditions in the code that look a bit complex, tried to explain as comments. Also, I'm not able to run tests locally and I can see from CI that something is wrong, maybe you can assist me with this?

aiday-mar commented:
Hi thanks for this PR. Today I am a bit busy, but I will make sure to look at this soon.

2mas committed:
fix test execution

2mas committed:
fix: remove unnecessary import and comment test suite meaning

2mas commented:
> Hi thanks for this PR. Today I am a bit busy, but I will make sure to look at this soon.

No stress at all. I got the test runner working now (not the .bat, but the integrated runner in vscode started working) so test suite is now repaired

aiday-mar committed:
polish

2mas commented:
Hi! Order of the initial selections are now lost since the list is already sorted from beginning, but since it requires an extra loop for it, it may not be super important, up to you. Probably its the cause for failing tests.

However the `continue` doesn't work that well, I'll show a small example

![image](https://github.com/user-attachments/assets/996e448c-bd42-488c-a6aa-21206ea7aa7e)
Moving down gets them unordered and selections are lost
![image](https://github.com/user-attachments/assets/d83a3abd-7063-4c0d-8df2-49f2bfe50ca5)

Would it be better to keep the `break` here, otherwise the nearest neighbour doesn't know that the top/bottom line was a no-op and takes its place?

--
I had some uncommitted code to remove the `movingMultipleLines` branching btw, it wasn't actually necessary since there was a break to abort the operation. I'll add it here if you find something useful to pick from it (also covers the case with multiple cursors on a single row only, they will disappear as it now):

```ts
public run(accessor: ServicesAccessor, editor: ICodeEditor): void {
		const languageConfigurationService = accessor.get(ILanguageConfigurationService);
		const autoIndent = editor.getOption(EditorOption.autoIndent);

		const selections = editor.getSelections() || [];
		const distinctSelectionsPerLine: Selection[] = [];
		const selectionsAfterLineMove: Selection[] = [];

		for (const selection of selections) {
			const startCol = selection.getStartPosition().column;
			const startLine = selection.getStartPosition().lineNumber;
			const newStartLine = this.down ? startLine + 1 : startLine - 1;

			const endCol = selection.getEndPosition().column;
			const endLine = selection.getEndPosition().lineNumber;
			const newEndLine = this.down ? endLine + 1 : endLine - 1;

			selectionsAfterLineMove.push(new Selection(newStartLine, startCol, newEndLine, endCol));
		}

		selections.sort((a, b) => a.endLineNumber - b.endLineNumber);
		let lastLineNumber = 0;

		for (const selection of selections) {
			if (lastLineNumber !== selection.endLineNumber) {
				distinctSelectionsPerLine.push(selection);
				lastLineNumber = selection.endLineNumber;
			}
		}

		if (this.down) {
			distinctSelectionsPerLine.reverse();
		}

		const model = editor.getModel();
		const lineCount = model !== null ? model.getLineCount() : 0;

		editor.pushUndoStop();

		for (const selection of distinctSelectionsPerLine) {
			if ((selection.startLineNumber === 1 && !this.down) || (selection.endLineNumber === lineCount && this.down)) {
				editor.pushUndoStop();
				return;
			}

			editor.executeCommand(this.id, new MoveLinesCommand(selection, this.down, autoIndent, languageConfigurationService));
		}

		editor.setSelections(selectionsAfterLineMove);
		editor.pushUndoStop();
	}
```

aiday-mar commented:
Hi @armanio123 thank you for the ping and apologies for this wait, there has been a lot of work as of late. The refactor I believe was necessary as we want PRs to adhere to a certain coding standard and be minimal and concise. I need to bring this PR up with my team and discuss it before merging.

aiday-mar commented:
Hi yes we need to address the functional differences too, thanks for bringing it up. The gaps need to be filled indeed.
